contracts/staking/ServiceStakingToken.sol

```diff
@@ -0,0 +1,199 @@
+    ///         Also, the final `require` statement is modified with this contract own `revert` statement.
+    /// @param token Token address.
+    /// @param to Address to transfer tokens to.
+    /// @param amount Token amount.
+    function safeTransfer(address token, address to, uint256 amount) internal {
+        bool success;
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+        // We'll write our calldata to this slot below, but restore it later.
+            let memPointer := mload(0x40)
+
+        // Write the abi-encoded calldata into memory, beginning with the function selector.
+            mstore(0, 0xa9059cbb00000000000000000000000000000000000000000000000000000000)
+            mstore(4, to) // Append the "to" argument.
+            mstore(36, amount) // Append the "amount" argument.
+
+            success := and(
+            // Set success to whether the call reverted, if not we check it either
+            // returned exactly 1 (can't just be non-zero data), or had no return data.
+                or(and(eq(mload(0), 1), gt(returndatasize(), 31)), iszero(returndatasize())),
+            // We use 68 because that's the total length of our calldata (4 + 32 * 2)
+            // Counterintuitively, this call() must be positioned after the or() in the
+            // surrounding and() because and() evaluates its arguments from right to left.
+                call(gas(), token, 0, 0, 68, 0, 32)
+            )
+
+            mstore(0x60, 0) // Restore the zero slot to zero.
+            mstore(0x40, memPointer) // Restore the memPointer.
+        }
+
+        if (!success) {
+            revert TransferFailed(token, address(this), to, amount);
+        }
+    }
+
+    /// @dev Checks token security deposit.
+    /// @param serviceId Service Id.
+    function _checkTokenSecurityDeposit(uint256 serviceId) internal view override {
+        // Get the service security token and deposit
+        (address token, uint96 securityDeposit) =
+            IServiceTokenUtility(serviceRegistryTokenUtility).mapServiceIdTokenDeposit(serviceId);
+
+        // The security token must match the contract token
+        if (securityToken != token) {
+            revert();
+        }
+
+        // The security deposit must be greater or equal to the minimum defined one
+        if (securityDeposit < minSecurityDeposit) {
+            revert();
+        }
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal override {
+        safeTransfer(securityToken, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    /// @dev Deposits funds for staking.
+    /// @param amount Token amount to deposit.
+    function deposit(uint256 amount) external {
+        // Distribute current staking rewards
+        _checkpoint(0);
```

Comment:
I am not sure it's sensible to do this call. Is it a must or did you add it for "convenience". In the later case I'd remove it.

Reply:
I can't see how else we can account correct rewards for all the services with the current balance that is going to change. We need to set a checkpoint for this current time, then increase the balance and calculate future rewards based on a new balance and starting with a new time.

Reply:
Why not? You can simply first update the balances? Anything that happened at t you can do at t+k.

Reply:
The balances can be updated only when calculating rewards for all the eligible services. To do this, need to traverse all of them and sync at the current timestamp.

Reply:
Otherwise we'll have to keep track of all the balance deposits and their times.

Reply:
we are talking about "balance" and "availableRewards", both of which can be updated without problem imo. Just don't see where in the code the issue is